core/trino-main/src/test/java/io/trino/metadata/TestMetadataListing.java

```diff
@@ -0,0 +1,175 @@
+import static io.trino.metadata.MetadataListing.listTablePrivileges;
+import static io.trino.metadata.MetadataListing.listTables;
+import static io.trino.metadata.MetadataListing.listViews;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestMetadataListing
+{
+    private static final String BROKEN_CATALOG = "broken_catalog";
+
+    private LocalQueryRunner queryRunner;
+    private Metadata metadata;
+    private AccessControl accessControl;
+
+    @BeforeClass
+    public void setUp()
+    {
+        queryRunner = LocalQueryRunner.create(TEST_SESSION);
+        MockConnectorFactory mockConnectorFactory = MockConnectorFactory.builder()
+                .withListSchemaNames(session -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withListTables((session, schema) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withListViews((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withGetViews((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withListMaterializedViews((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withGetMaterializedViews((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withListTablePrivileges((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .withStreamTableColumns((session, prefix) -> {
+                    throw new TrinoException(GENERIC_INTERNAL_ERROR, "Catalog is broken");
+                })
+                .build();
+        queryRunner.createCatalog(BROKEN_CATALOG, mockConnectorFactory, ImmutableMap.of());
+        metadata = queryRunner.getMetadata();
+        accessControl = queryRunner.getAccessControl();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        try {
+            if (queryRunner != null) {
+                queryRunner.close();
+            }
+        }
+        finally {
+            metadata = null;
+            accessControl = null;
+            queryRunner = null;
+        }
+    }
+
+    @Test
+    public void testListSchemasExceptionHandling()
+    {
+        Session session = createSession();
+        assertThatThrownBy(() -> listSchemas(session, metadata, accessControl, BROKEN_CATALOG))
```

Comment:
Can you run regular query instead and use `QueryAssertions`. The thing that this method is called for query is kind of implementation detail and we would like to have test for UX that when user run a query against metadata they got information which catalog is broken.